### PR TITLE
Move restricted/pure-eval access control out of the evaluator and into the accessor

### PR DESCRIPTION
**Pull request by edolstra**
# Motivation

Extracted from lazy-trees.

# Context

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Comment by Ericson2314** [comment]
Didn't read all the code while it is still draft, but the concept is quite nice!

File: src/libfetchers/filtering-input-accessor.hh [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+ * A function that should throw an exception of type
+ * `RestrictedPathError` explaining that access to `path` is
+ * forbidden.
+ */
+typedef std::function<RestrictedPathError(const CanonPath & path)> MakeNotAllowedError;
+
+/**
+ * An abstract wrapping `InputAccessor` that performs access
+ * control. Subclasses should override `checkAccess()` to implement an
+ * access control policy.
```
Comment (roberth) [inline src/libfetchers/filtering-input-accessor.hh:17]:
```suggestion
 * control. Subclasses should override `isAllowed()` to implement an
 * access control policy. The error message is customized at construction.
```

File: tests/functional/restricted.sh [hunk 2]
```diff
@@ -14,8 +14,8 @@ nix-instantiate --restrict-eval --eval -E 'builtins.readFile ./simple.nix' -I sr
 (! nix-instantiate --restrict-eval --eval -E 'builtins.readDir ../../src/nix-channel')
 nix-instantiate --restrict-eval --eval -E 'builtins.readDir ../../src/nix-channel' -I src=../../src
 
-(! nix-instantiate --restrict-eval --eval -E 'let __nixPath = [ { prefix = "foo"; path = ./.; } ]; in <foo>')
-nix-instantiate --restrict-eval --eval -E 'let __nixPath = [ { prefix = "foo"; path = ./.; } ]; in <foo>' -I src=.
+(! nix-instantiate --restrict-eval --eval -E 'let __nixPath = [ { prefix = "foo"; path = ./.; } ]; in builtins.readFile <foo/simple.nix>')
+nix-instantiate --restrict-eval --eval -E 'let __nixPath = [ { prefix = "foo"; path = ./.; } ]; in builtins.readFile <foo/simple.nix>' -I src=.
```
Comment (roberth) [inline tests/functional/restricted.sh:18]:
```suggestion
expect 1 nix-instantiate --restrict-eval --eval -E 'let __nixPath = [ { prefix = "foo"; path = ./.; } ]; in builtins.readFile <foo/simple.nix>' 2>&1 | grep "restricted"
nix-instantiate --restrict-eval --eval -E 'let __nixPath = [ { prefix = "foo"; path = ./.; } ]; in builtins.readFile <foo/simple.nix>' -I src=.
```

Make sure it's the right error.

File: tests/functional/restricted.sh
Comment (roberth) [inline tests/functional/restricted.sh:1]: We have a regression with symlinks. Needs a test case.
Currently:

```console
ls -l tunnel.d/
total 4
lrwxrwxrwx 1 user user 2 Nov 30 23:46 tunnel -> ..

$ ./result/bin/nix-instantiate --restrict-eval --eval -E 'let __nixPath = [ { prefix = "foo"; path = ./tunnel.d; } ]; in builtins.readFile <foo/tunnel/aliens-truth>' -I tunnel.d
"They're among us.\n"
```

**Review by roberth — CHANGES_REQUESTED** [review]
Symlinks strike again.
Otherwise looks like the right direction.

File: src/libexpr/primops.cc [hunk 3]
```diff
@@ -2178,60 +2160,65 @@ static RegisterPrimOp primop_toFile({
...
-        // be rewritten to the actual output).
-        auto rewrites = state.realiseContext(context);
-        path = state.toRealPath(rewriteStrings(path, rewrites), context);
-
         StorePathSet refs;
 
-        if (state.store->isInStore(path)) {
+        if (path.accessor == state.rootFS && state.store->isInStore(path.path.abs())) {
+            // FIXME: handle CA derivation outputs (where path needs to
+            // be rewritten to the actual output).
```
[the remark this answers is not in the document]
Comment (edolstra) [reply src/libexpr/primops.cc:2203]: No, it's a moved comment.

File: tests/functional/restricted.sh (hunk 2 above)
Comment (edolstra) [reply tests/functional/restricted.sh]: Done.

File: src/libfetchers/filtering-input-accessor.hh [hunk 4]
```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+#include "input-accessor.hh"
+
+namespace nix {
+
+/**
+ * A function that should throw an exception of type
+ * `RestrictedPathError` explaining that access to `path` is
+ * forbidden.
```
[the remark this answers is not in the document]
Comment (edolstra) [reply src/libfetchers/filtering-input-accessor.hh:10]: No it should really say that it's required to throw `RestrictedPathError`.

[the remark this answers is not in the document]
Comment (roberth) [reply src/libfetchers/filtering-input-accessor.hh:10]: That's also ok, but then it should be 

```c++
typedef std::function<void(const CanonPath & path)> ThrowNotAllowedError;
```

**Review by roberth — APPROVED** [review]

**Comment by infinisil** [comment]
This caused a regression: https://github.com/NixOS/nix/issues/9901

**Pull request merged**